xbmc/platform/darwin/tvos/filesystem/TVOSFile.cpp

```diff
@@ -0,0 +1,274 @@
+      return 0;
+    }
+  }
+  errno = ENOENT;
+  return -1;
+}
+
+bool CTVOSFile::Open(const CURL& url)
+{
+  if (CDarwinNSUserDefaults::KeyFromPathExists(url.Get()))
+  {
+    m_url = url;
+    m_position = 0;
+    CacheStat(url, &m_cachedStat);
+    return true;
+  }
+  else
+  {
+    // fallback to posixfile
+    m_pFallbackFile = new CPosixFile();
+    return m_pFallbackFile->Open(url);
+  }
+}
+
+bool CTVOSFile::OpenForWrite(const CURL& url, bool bOverWrite /* = false */)
+{
+  if (CDarwinNSUserDefaults::KeyFromPathExists(url.Get()) && !bOverWrite)
+    return false; // no overwrite
+
+  bool ret = WantsFile(url); // if we want the file we can write it ...
+  if (ret)
+  {
+    m_url = url;
+    m_position = 0;
+  }
+  return ret;
+}
+
+bool CTVOSFile::Delete(const CURL& url)
+{
+  bool ret = CDarwinNSUserDefaults::DeleteKeyFromPath(url.Get(), true);
+
+  if (!ret)
+  {
+    CPosixFile posix;
+    ret = posix.Delete(url);
+  }
+  return ret;
+}
+
+bool CTVOSFile::Exists(const CURL& url)
+{
+  bool ret = CDarwinNSUserDefaults::KeyFromPathExists(url.Get());
+  if (!ret)
+  {
+    CPosixFile posix;
+    ret = posix.Exists(url);
+  }
+  return ret;
+}
+
+int CTVOSFile::Stat(const CURL& url, struct __stat64* buffer)
+{
+  int ret = CacheStat(url, buffer);
+  if (ret < 0)
+  {
+    CPosixFile posix;
+    ret = posix.Stat(url, buffer);
+  }
+  return ret;
+}
+
+bool CTVOSFile::Rename(const CURL& url, const CURL& urlnew)
+{
+  bool ret = false;
+  if (Exists(url) && !Exists(urlnew) && WantsFile(urlnew))
+  {
+    void* lpBuf = nullptr;
+    size_t uiBufSize = 0;
+    if (CDarwinNSUserDefaults::GetKeyDataFromPath(url.Get(), lpBuf,
+                                                  uiBufSize)) // get size from old file
+    {
+      lpBuf = (void*)new char[uiBufSize];
+      if (CDarwinNSUserDefaults::GetKeyDataFromPath(url.Get(), lpBuf, uiBufSize)) // read old file
+      {
+        if (CDarwinNSUserDefaults::SetKeyDataFromPath(urlnew.Get(), lpBuf, uiBufSize,
+                                                      true)) // write to new url
+        {
+          // remove old file
+          Delete(url);
+          ret = true;
+        }
+      }
+      delete[] static_cast<char*>(lpBuf);
+    }
+  }
+
+  if (!ret)
+  {
+    CPosixFile posix;
+    ret = posix.Rename(url, urlnew);
+  }
+  return ret;
+}
+
+int CTVOSFile::Stat(struct __stat64* buffer)
+{
+  memcpy(buffer, &m_cachedStat, sizeof(struct __stat64));
+  return 0;
+}
+
+ssize_t CTVOSFile::Read(void* lpBuf, size_t uiBufSize)
+{
+  if (m_pFallbackFile != nullptr)
+    return m_pFallbackFile->Read(lpBuf, uiBufSize);
+
+  void* lpBufInternal = nullptr;
+
+  if (m_position > 0 && m_position == GetLength())
+    return 0; // simulate read 0 bytes on EOF
+
+  if (CDarwinNSUserDefaults::GetKeyDataFromPath(m_url.Get(), lpBufInternal,
+                                                uiBufSize)) // read size of file
+  {
+    lpBufInternal = static_cast<void*>(new char[uiBufSize]);
```

Comment:
use smart pointer

Reply:
discussed this one with you. At this stage i'll leave it as is. May be able to change it with some refactoring. Will look into it at a later date.